oda_api/api.py

```diff
@@ -1339,3 +1340,48 @@ def from_response_json(cls, res_json, instrument, product):
                 p.meta_data = p.meta
 
         return d
+
+class ProgressReporter(object):
+    """
+    The class allows to report task progress to end user
+    """
+    def __init__(self):
+        self._callback = None
+        callback_file = ".oda_api_callback"  # perhaps it would be better to define this constant in a common lib
+        if not os.path.isfile(callback_file):
+            return
+        with open(callback_file, 'r') as file:
+            self._callback = file.read().strip()
+
+    @property
+    def enabled(self):
+        return self._callback is not None
+
+    def report_progress(self, stage: str=None, progress: int=50, substage: str=None, subprogress: int=None, message:str=None):
+        """
+        Report progress via callback URL
+        :param stage: current stage description string
+        :param stage: current stage progress in %
+        :param stage: current substage description string
+        :param stage: current substage progress in %
+        :param message: message to pass
+        """
+        callback_payload = {k: str(v) for k, v in locals().items() if v is not None and k != 'self'}
```

Comment:
this is going to put the entire content of locals as a payload for the request, is it really needed? can it be filtered to remove not-needed items?

Reply:
I think it's ok, as locals() in this content are only method parameters

Reply:
I think it's ok, as locals() in this content are only method parameters

Reply:
> this is going to put the entire content of locals as a payload for the request, is it really needed? can it be filtered to remove not-needed items?

I agree with @burnout87 that this kind of code-shortening trickery, although inventive, is prone to mistakes - what if someone adds any simple assignment, in some check?
You do spell out all of the parameters twice - in function spec and in the doc string. Adding one more time does not hurt here.

Or, if you want shorter, use kwargs?
Or, inspect function parameters? 

But really, it's not too much of a duplication to introduce error-prone complexity.
Even if I indeed unironically appreciate inventive code!